design/keycloak.x/storage-hot-rod.md

```diff
@@ -0,0 +1,292 @@
+
+### `TransactionManager` lookup
+
+The Hot Rod client is able to automatically lookup `TransactionManager` using [`GenericTransactionManagerLookup`](https://docs.jboss.org/infinispan/9.4/apidocs/org/infinispan/transaction/lookup/GenericTransactionManagerLookup.html) when running in Java EE application server. When there is no `TransactionManager` running in the container, [`RemoteTransactionManager`](https://docs.jboss.org/infinispan/10.0/apidocs/org/infinispan/client/hotrod/transaction/manager/RemoteTransactionManager.html) is returned which can be enlisted to [KeycloakTransactionManager](https://github.com/keycloak/keycloak/blob/af3b573d196af882dfb25cdccb98361746e85481/server-spi/src/main/java/org/keycloak/models/KeycloakTransactionManager.java). 
+
+### Hot Rod client transaction mode
+
+The Hot Rod client transaction can work in the following modes:
+- **NONE** - non-transactional - when using this setting, we would not be able to achieve point number 3 from our requirements.
+- **NON_XA** - in this mode, the `RemoteCache` interacts with the `TransactionManager` via [`Synchronization`](https://docs.oracle.com/javaee/7/api/javax/transaction/Synchronization.html); this means the changes are propagated to the Infinispan storage only when the main transaction successfully commits; however, in the case of a failure during storing data in the Infinispan server the main transaction is not rolled back. 
+- **NON_DURABLE_XA** - `RemoteCache` interacts with the `TransactionManager` using [`XAResource`](https://docs.oracle.com/javaee/7/api/javax/transaction/xa/XAResource.html) without possibility of [recovery](https://infinispan.org/docs/stable/titles/developing/developing.html#tx_recovery).
+- **FULL_XA** - `RemoteCache` interacts with the `TransactionManager` using [`XAResource`](https://docs.oracle.com/javaee/7/api/javax/transaction/xa/XAResource.html) with the possibility of recovery. Should be avoided due to the [documentation note](https://infinispan.org/docs/stable/titles/hotrod_java/hotrod_java.html#hr_transactions_config_server) about performance degradation.
+
+Based on the above, **NON_DURABLE_XA** seems to be the best option to use given the requirements.
+
+# Infinispan Initialization
+
+## Remote caches
+
+Configuration of the remote caches will be stored declaratively in a xml file. The caches can be created either:
+* by a user when starting the Infinispan server using the provided configuration
+* by Keycloak which will check if the caches exist on the Infinispan server during the startup and if not, it will create them. However some users may want to tweak the configuration based on their special needs in which case they will create the caches themselves.
```

Comment:
```suggestion
* For developers, by Keycloak which will check if the caches exist on the Infinispan server during the startup and if not, it will create them. However some users may want to tweak the configuration based on their special needs in which case they will create the caches themselves.
```